### ISSUE TITLE
c++ compilation creates namespace for uninstantiated functor

### ISSUE THREAD

**Issue by tjhance** [secure-foundations/dafny#44]
```
abstract module Ifc { }

abstract module InputOutputIfc refines Ifc { }

abstract module DiskSSM(IOIfc: InputOutputIfc) { }

module DiskSSMTokens(ioifc: InputOutputIfc, ssm: DiskSSM(ioifc)) {
  type Token = int
}

abstract module AIOParams { }

module AIO(aioparams: AIOParams, ioifc: InputOutputIfc, ssm: DiskSSM(ioifc)) {
  import T = DiskSSMTokens(ioifc, ssm)
}
```

Produces the undesirable namespace:

```
namespace DiskSSMTokens_1_ON_ioifc_ssm__Compile  {
```

**Comment by parno** [comment]
Can you say more about the problem this creates?

**Comment by tjhance** [comment]
In some cases the invalid namespace has code in it that references other namespaces that (rightly) don't exist. So this leads to C++ compile errors.

DiskSSMTokens(iofic, ssm) isn't a real module. In fact, I wouldn't expect this Dafny program to produce _any_ c++ code, since there are not actually any instantiated modules.

if I were to actually instantiate AIO with, say, `AIO(A, B, C)` (where `A`, `B`, and `C` are concrete modules), that would in turn cause an instantiation `DiskSSMTokens(B, C)`  and so on, so I'd expect to find a compiled namespace something like `DiskSSMTokens_1_ON_B_C__Compile`. But if there's no instantiation, there shouldn't be any code.

**Comment by parno** [comment]
Should be fixed by 3da47d2
